### PR TITLE
fix: empty datanode command for loading historical data into pre-populated node

### LINKED ISSUE #7152: Loading historical data requires easy way to empty the node
3 primary use cases for dehistory:

1) Get node up to latest height and processing new blocks
2) Get a node that already contains data up to the latest height (filling in the gap) and processing new blocks
3) Load all (or some) historical data 

Cases 1 and 2 are handled by the autoinitialise flag, no manual intervention required.  Case 3 however requires that the runners are archival nodes carry out the following process:

A) start datanode (with autoinitialise=true) - this will get the datanode up to the networks current height and start producing segments locally
B) Fetch the required historical asynchronously - as datanode is running it will not get behind during this process
C) Stop the datanode, empty the datanode and load all data from dehistory.

Currently there is no easy way, without connecting to DB and executing some drop database etc commands, to empty the datanode.  Automating the dropping of the data in the case where the history to load is older than that in the datanode would be possible but not desirable, better that the operator takes this decision explicitly. - The error message when load fails due to this scenario is clear e.g.

`loading history from height 1 is not possible as it is before the datanodes oldest block height 19001, to load this history first empty the datanode`

### PR DESCRIPTION
**Pull request by ettec**
closes #7152

### CODE REVIEW

**Comment by wwestgarth** [comment]
For Scenario 3, is that not what the `MinimumBlockCount = 1` config option to auto-initialise is for?

File: cmd/data-node/commands/dehistory/empty-datanode.go [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+	err = paths.ReadStructuredFile(configFilePath, &cmd.Config)
+	if err != nil {
+		return fmt.Errorf("failed to read configuration:%w", err)
+	}
+
+	if datanodeLive(cmd.Config) {
+		return fmt.Errorf("datanode must be shutdown before datanode can be emptied")
+	}
+
+	yes := flags.YesOrNo("This will remove all data from datanode, are you sure?")
```
Comment (wwestgarth) [inline cmd/data-node/commands/dehistory/empty-datanode.go:44]: We're going to need a `--force` option or something that circumvents the `yes/no` so people can script against it.

File: cmd/data-node/commands/dehistory/empty-datanode.go [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+	)
+	defer log.AtExit()
+
+	vegaPaths := paths.New(cmd.VegaHome)
+	configFilePath, err := vegaPaths.CreateConfigPathFor(paths.DataNodeDefaultConfigFile)
+	if err != nil {
+		return fmt.Errorf("couldn't get path for %s: %w", paths.DataNodeDefaultConfigFile, err)
+	}
+
+	err = paths.ReadStructuredFile(configFilePath, &cmd.Config)
```
Comment (wwestgarth) [inline cmd/data-node/commands/dehistory/empty-datanode.go:35]: As in https://github.com/vegaprotocol/vega/issues/7028 can we make sure `config.Config` can be overriden by CLI options?

**Comment by ettec** [comment]
> For Scenario 3, is that not what the `MinimumBlockCount = N` config option to auto-initialise is for? We keep fetching segments until we have it all?

That works when the history is small, as its fetched synchronously, thus the network is moving on while the fetch is happening, but thats ok if we're talking seconds/minutes.  Where the history is large (weeks/months whatever, i.e. setting up a new archive node) by the time the history is fetched the node would again be significantly behind the current network height, so for example,  there would be no suitable core snapshot available to start from.  Hence why the fetch of large sections of history has to be done async.

File: cmd/data-node/commands/dehistory/empty-datanode.go (hunk 2 above)
Comment (ettec) [reply cmd/data-node/commands/dehistory/empty-datanode.go:35]: phils in the middle of that, some code I will need to pickup, gonna wait till he merges

**Comment by wwestgarth** [comment]
> > For Scenario 3, is that not what the `MinimumBlockCount = N` config option to auto-initialise is for? We keep fetching segments until we have it all?
> 
> That works when the history is small, as its fetched synchronously, thus the network is moving on while the fetch is happening, but thats ok if we're talking seconds/minutes. Where the history is large (weeks/months whatever, i.e. setting up a new archive node) by the time the history is fetched the node would again be significantly behind the current network height, so for example, there would be no suitable core snapshot available to start from. Hence why the fetch of large sections of history has to be done async.

Oh ok, so this would work:
```
vega datanode dehistory fetch EVERYTHING

# takes some time

vega datanode start # let autoinitiliase fill in the rest that appear while everything was fetching earlier
```

but it just means you won't have a datanode up while you're waiting for a big fetch?

I'm wondering then, whether it should just be an option to `dehistory load` instead of something separate? eg.
```
vega datanode dehistory load # get some error because you have stuff already in the db, so you do

vega datanode dehistory load --wipe-existing
```
?

**Comment by ettec** [comment]
> > > For Scenario 3, is that not what the `MinimumBlockCount = N` config option to auto-initialise is for? We keep fetching segments until we have it all?
> > 
> > 
> > That works when the history is small, as its fetched synchronously, thus the network is moving on while the fetch is happening, but thats ok if we're talking seconds/minutes. Where the history is large (weeks/months whatever, i.e. setting up a new archive node) by the time the history is fetched the node would again be significantly behind the current network height, so for example, there would be no suitable core snapshot available to start from. Hence why the fetch of large sections of history has to be done async.
> 
> Oh ok, so this would work:
> 
> ```
> vega datanode dehistory fetch EVERYTHING
> 
> # takes some time
> 
> vega datanode dehistory load
> vega datanode start # let autoinitiliase fill in the rest that appeared while everything was fetching earlier
> ```
> 
> but it just means you won't have a datanode up while you're waiting for a big fetch?
> 
> I'm wondering then, whether it should just be an option to `dehistory load` instead of something separate? eg.
> 
> ```
> vega datanode dehistory load # get some error because you have stuff already in the db, so you do
> 
> vega datanode dehistory load --wipe-existing-or-something
> ```
> 
> ?

possibly that would work but with a large amount of data the autoinitialise could then take a long time to auto init and you are back to square one as by the time autoinit has pulled the data you would be a long way behind current height

**Comment by ettec** [comment]
@wwestgarth closing this - I think your suggestion to add a '--wipe-existing' option to the load command is a better solution, going to close this PR and do that instead.

**Pull request closed without merging**